### PR TITLE
Optimize internal cache

### PR DESCRIPTION
**Pull request by vemv**
> Fixes https://github.com/clojure-emacs/orchard/issues/186

Replaces the atom cache that would easily grow to huge values with a `LruCache` with capacity 250.

We also mostly stop populating the cache in advance. It's fairly pointless if we use a bounded capacity - no random classes are more important than others.

(We do populate the cache for a few key frequently-used classes, and whatever underlying classes they also need)

Cheers - V

### CODE REVIEW

File: src/orchard/java.clj [hunk 1]
```diff
@@ -223,14 +225,16 @@
 ;; and recompiled, we cache such classes with last-modified property, so that we
 ;; know when to purge those classes from cache.
 
-(def cache (atom {}))
+;; 250 is large enough to be useful (and hold a few key classes, like Object),
+;; and small enough to not incur into OOMs.
+(def ^Map cache (LruMap. 250))
```
Comment (bbatsov) [inline src/orchard/java.clj:230]: I'm guessing that even double the size it wouldn't be an issue. You might want to expand the documentation a little bit to elaborate that this cache size is bounded and that's the reason why we went with a custom implementation.

**Comment by bbatsov** [comment]
Overall it looks OK to me. Generally I avoid introducing Java code when possible, but in this case it's fairly simple so not a big deal. Alternatively you could have done something like https://github.com/clojure/core.cache/blob/master/src/main/clojure/clojure/core/cache.clj#L211

File: java/mx/cider/orchard/LruMap.java [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+import java.util.Collection;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.concurrent.locks.ReadWriteLock;
+
+/**
+ * A <code>java.util.Map</code> with a capacity expressed as <code>max_size</code>
```
Comment (bbatsov) [inline java/mx/cider/orchard/LruMap.java:11]: You might mention that LRU stands for least recently used. :-) It took me a moment to remember this when I saw the class name.

**Comment by vemv** [comment]
Thanks for the review!

**Pull request merged**